ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,525 @@
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->fragment.first != NULL && uriparser_uri->fragment.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->fragment.first, uriparser_uri->fragment.afterLast - uriparser_uri->fragment.first);
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static void *uriparser_malloc(UriMemoryManager *memory_manager, size_t size)
+{
+	return emalloc(size);
+}
+
+static void *uriparser_calloc(UriMemoryManager *memory_manager, size_t nmemb, size_t size)
+{
+	return ecalloc(nmemb, size);
+}
+
+static void *uriparser_realloc(UriMemoryManager *memory_manager, void *ptr, size_t size)
+{
+	return realloc(ptr, size);
+}
+
+static void *uriparser_reallocarray(UriMemoryManager *memory_manager, void *ptr, size_t nmemb, size_t size)
+{
+	const size_t total_size = nmemb * size;
+
+	/* check for unsigned overflow */
+	if ((nmemb != 0) && (total_size / nmemb != size)) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	return erealloc(ptr, total_size);
+}
+
+static void uriparser_free(UriMemoryManager *memory_manager, void *ptr)
+{
+	efree(ptr);
+}
+
+zend_result uriparser_request_init(void)
+{
+	uriparser_memory_manager.calloc = uriparser_calloc;
+	uriparser_memory_manager.malloc = uriparser_malloc;
+	uriparser_memory_manager.realloc = uriparser_realloc;
+	uriparser_memory_manager.reallocarray = uriparser_reallocarray;
+	uriparser_memory_manager.free = uriparser_free;
+
+	zend_hash_init(&uriparser_property_handlers, 8, NULL, NULL, true);
+
+	return SUCCESS;
+}
+
+static uriparser_uris_t *uriparser_create_uris(
+	UriUriA *uri, zend_string *uri_str,
+	UriUriA *normalized_uri, zend_string *normalized_uri_str
+) {
+	uriparser_uris_t *uriparser_uris = emalloc(sizeof(uriparser_uris_t));
+
+	uriparser_uris->uri = uri;
+	uriparser_uris->uri_str = uri_str;
+	uriparser_uris->normalized_uri = normalized_uri;
+	uriparser_uris->normalized_uri_str = normalized_uri_str;
+
+	return uriparser_uris;
+}
+
+static void throw_invalid_uri_exception(void)
+{
+	zval exception;
+
+	object_init_ex(&exception, uri_invalid_uri_exception_ce);
+
+	zval value;
+	ZVAL_STRING(&value, "URI parsing failed");
+	zend_update_property_ex(uri_whatwg_invalid_url_exception_ce, Z_OBJ(exception), ZSTR_KNOWN(ZEND_STR_MESSAGE), &value);
+	zval_ptr_dtor_str(&value);
+
+	zend_throw_exception_object(&exception);
+}
+
+void *uriparser_parse_uri_ex(const zend_string *uri_str, const uriparser_uris_t *uriparser_base_urls, bool silent)
+{
+	UriUriA *uriparser_uri = emalloc(sizeof(UriUriA));
+
+	/* uriparser keeps the originally passed in string, while lexbor may allocate a new one. */
+	zend_string *original_uri_str = zend_string_init(ZSTR_VAL(uri_str), ZSTR_LEN(uri_str), false);
+	if (ZSTR_LEN(original_uri_str) == 0 ||
+		uriParseSingleUriExMmA(uriparser_uri, ZSTR_VAL(original_uri_str), ZSTR_VAL(original_uri_str) + ZSTR_LEN(original_uri_str), NULL, &uriparser_memory_manager) != URI_SUCCESS
```

Comment:
Passing the memory manager to all functions is kinda error-prone. Is it possible to set it globally somehow?

Reply:
Yes, I can confirm it's error prone. A few weeks ago I lost a few hours because the memory manager was forgotten to be passed to the freeing function.

So far I didn't realize it was possible to override the functions of the default manager (https://github.com/php/php-src/blob/d585a5609d9d9ad9390eeb8c3109b64ff02bd632/ext/uri/uriparser/src/UriMemory.c#L464), but it's certainly possible. Thanks for the idea!